chore: use tagged literals in tests

grigoriy-reshetniak opened the pull request:
<!-- PR title should follow Angular Commit Message Conventions (https://github.com/angular/angular.js/blob/master/DEVELOPERS.md#-git-commit-guidelines) -->
https://github.com/KaTeX/KaTeX/pull/3902 looks abandoned, so I've decided to try to finish it.

**What is the previous behavior before this PR?**
`expect("\\mathinner{\\langle{\\psi}\\rangle}").toParse();`

**What is the new behavior after this PR?**
``expect`\mathinner{\langle{\psi}\rangle}`.toParse();``

<!-- If this PR contains a breaking change, please uncomment following line -->
<!-- BREAKING CHANGE: describe its impact and migration methods -->

<!-- If this PR fixes or closes issues, please uncomment following line and change the issue number -->
<!-- Fixes #1, Closes #2 -->
Fixes https://github.com/KaTeX/KaTeX/issues/1717

grigoriy-reshetniak commented:
@edemaine could you please review and start CI job? Tests are passing locally.
![image](https://github.com/KaTeX/KaTeX/assets/8845364/6ab63752-c598-4f0e-a9a3-5c3b5fc70982)

edemaine commented on test/katex-spec.js:
```diff
@@ -1671,42 +1670,42 @@ describe("A font parser", function() {
-        expect("\\pmb{\\frac{x^2}{x_1}}").toBuild();
-        expect("\\pmb{}").toBuild();
-        expect("\\def\\x{1}\\pmb{\\x\\def\\x{2}}").toParseLike("\\pmb{1}");
+        expect`\pmb{\\mu}`.toBuild();
```
```suggestion
        expect`\pmb{\mu}`.toBuild();
```

edemaine commented on test/katex-spec.js:
```diff
@@ -1671,42 +1670,42 @@ describe("A font parser", function() {
+        expect`\pmb{\\mu}`.toBuild();
+        expect`\pmb{=}`.toBuild();
+        expect`\pmb{+}`.toBuild();
+        expect`\pmb{\\frac{x^2}{x_1}}`.toBuild();
```
```suggestion
        expect`\pmb{\frac{x^2}{x_1}}`.toBuild();
```

edemaine commented on test/katex-spec.js:
```diff
@@ -1523,18 +1522,18 @@ describe("A TeX-compliant parser", function() {
+        expect`\oiiint_i^n`.toBuild();
+        expect`\oint\nolimits_i^n`.toBuild();
+        expect`\oiint\nolimits_i^n`.toBuild();
+        expect`\oiiint\nolimits_i^n`.toBuild();
```
A lot of these tests were buggy before with `\n`! Nice to fix them.

edemaine commented on test/katex-spec.js:
```diff
@@ -3410,27 +3409,27 @@ describe("A macro expander", function() {
+        expect("\\char`\\%").toParseLike`\char'45`;
+        expect("\\char`\\%").toParseLike`\char"25`;
+        expect`\char`.not.toParse();
+        expect`\char\``.not.toParse();
```
Does this convert to `"\\char`"` or `"\\char\\`"`? I'm not sure...

edemaine commented on test/katex-spec.js:
```diff
@@ -811,15 +811,15 @@ describe("A text parser", function() {
-        expect("\\text{\\sect}\\sdot\\sub\\sube\\supe").toBuild();
-        expect("\\Tau\\thetasym\\weierp\\Zeta").toBuild();
+        expect`\lang\N\darr\R\dArr\Z\Darr\alef\rang`.toBuild();
+        expect`\alefsym\uarr\Alpha\uArr\Beta\Uarr\Chi`.toBuild();
```
I thought `\u` didn't work in raw strings because it's a Unicode escape. See e.g.

https://github.com/KaTeX/KaTeX/blob/bedce93260cc413abb7730a5693dc27222297178/test/katex-spec.js#L2919

Is that incorrect? (It might be; I'm not sure.)

grigoriy-reshetniak replied:
No, you right. I'm changing it to double backslash 
![image](https://github.com/KaTeX/KaTeX/assets/8845364/9144412b-198b-4123-adb5-8f088c84bd8e)

edemaine replied:
This isn't the right experiment though. I think ``String.raw`...` `` is closer, but I couldn't immediately find how we're providing template strings creation functions.

grigoriy-reshetniak commented:
I have fixed 2 issues with double backslashes and reverted every test that includes `\u` or backtick. While backtick test passes,  the string that expect resolves to has changed. Some test with `\u`, like `An under-accent parser` work fine with backticks and single backslash, but I'm not sure why  and I better move it to separate PR after I figure it out.

edemaine commented:
I found where we handle tagged literals:

https://github.com/KaTeX/KaTeX/blob/bedce93260cc413abb7730a5693dc27222297178/test/helpers.js#L17-L20

This gets called e.g. in `getBuilt`. So I think the correct test is:

```js
const r = x => x != null && x.hasOwnProperty('raw') ? x.raw[0] : x;
r`\url`
```

which correctly generates `"\\url"`. So I think all the `\u` tests can in fact be restored to use backticks. Sorry for the runaround!

Maybe you can update these comments too:

https://github.com/KaTeX/KaTeX/blob/bedce93260cc413abb7730a5693dc27222297178/test/katex-spec.js#L2919

https://github.com/KaTeX/KaTeX/blob/bedce93260cc413abb7730a5693dc27222297178/docs/api.md#L16

I think what is still problematic is `${` (because of interpolations, and `\$` won't fix it) and `` ` `` that isn't preceded by a backslash (again `` \` `` would produce something different than `` ` ``). But otherwise things template strings should be OK: `\u` and `\x` do not seem to pose any issue *when using `String.raw` (or `r` as in our tests)*.

In the future, can you add commits instead of rebasing? Otherwise I can't really review the new changes. We squash commit when accepting the PR, so the commit history doesn't matter except for review.

edemaine reviewed (approved):
Thank you!
